Variationally consistent mortar friction

recuero opened the pull request:
This formulation of mortar frictional forces:

- Uses a tangential vector field from element faces (for now)
- Does not have a fully objective definition of relative velocity (probably still okay for many applications). See Ref. [2] of #17495 for details.
- ~~Not usable for general stick-slip behavior (it doesn't converge).  Strategies to alleviate this issue to be implemented~~.
- Implements initial primal-dual active set strategy 

For reason, design, and impact see  #17495.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/716736/) on 9f05d10 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17497/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17497/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 36c8f58721fa53518807f96a46b703998c7ad311`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/716762/) on 1626733 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17497/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17497/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c9c0f8fc799755215f0db5c9507f3441e2f0fe45`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/716787/) on d22b498 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17497/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17497/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c9c0f8fc799755215f0db5c9507f3441e2f0fe45`

recuero commented on framework/src/interfaces/NeighborCoupleable.C:
```diff
@@ -60,6 +60,23 @@ NeighborCoupleable::adCoupledNeighborValue(const std::string & var_name, unsigne
+    mooseError(
+        "adCoupledNeighborValueDot returns a data structure with derivatives. Explicit schemes "
+        "use old solution data which do not have derivatives so adCoupledNeighborValueDot is "
+        "not appropriate. Please use coupledNeighborValue instead");
```
Use `coupledNeighborValueDot`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/716822/) on 0e54411 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17497/site)

This comment will be updated on new commits.

recuero commented on modules/contact/src/constraints/ApplyFrictionConstraintLMMechanicalContact.C:
```diff
@@ -0,0 +1,295 @@
+
+      Real integrated_velocity = _node_to_integrated_velocity[_current_node];
+      LocalDN v_dot_tan(integrated_velocity);
+      //
+      //      (*_jacobian).
+      //      Moose::derivInsert(
+      //          v_dot_tan.derivatives(),
+      //          0,
+      //          (*_jacobian)(_current_node->dof_number(0, _var.number(), 0),
+      //          _connected_dof_indices[_j]));
+      //      Moose::derivInsert(
+      //          v_dot_tan.derivatives(),
+      //          1,
+      //          (*_jacobian)(_current_node->dof_number(0, _var.number(), 0),
+      //          _connected_dof_indices[_j]));
+      //      Moose::derivInsert(
+      //          v_dot_tan.derivatives(),
+      //          2,
+      //          (*_jacobian)(_current_node->dof_number(0, _var.number(), 0),
+      //          _connected_dof_indices[_j]));
```
@lindsayad This commented out code was triggering an uninitialized read error in valgrind after the block had been in contact for a while and it was breaking off contact (even though these derivative additions do help convergence). Would you please review when you get the chance?

recuero commented on modules/contact/test/tests/bouncing-block-contact_pdass/try_gitterle.i:
```diff
@@ -0,0 +1,274 @@
+    use_displaced_mesh = true
+  []
+
+ # Create gap_variable
+  [weighted_gap_indicator]
+    type = ComputeWeightedGapLMMechanicalContactIndicator
+    primary_boundary = 20
+    secondary_boundary = 10
+    primary_subdomain = 4
+    secondary_subdomain = 3
+    variable = gap_variable
+    disp_x = disp_x
+    disp_y = disp_y
+    use_displaced_mesh = true
+  []
+
+ # Create vel_variable
+  [tangential_vel]
+    type = ComputeWeightedTangentialVelocityLMMechanicalContactIndicator
+    primary_boundary = 20
+    secondary_boundary = 10
+    primary_subdomain = 4
+    secondary_subdomain = 3
+    variable = vel_variable
+    disp_x = disp_x
+    disp_y = disp_y
+    use_displaced_mesh = true
+    contact_pressure = normal_lm
+  []
```
@lindsayad  Here, I am using these "redundant" computations of weighted gap and tangential velocities. Just to be used to define the actual frictional constraints based on PDASS.
I guess one drawback of doing this is losing derivative information coming from the the mortar constraint. You mentioned you had an alternative way for doing this based on an object that loops over all the mortar segments contributing to a secondary node.

recuero commented on modules/contact/test/tests/bouncing-block-contact_pdass/try_gitterle.i:
```diff
@@ -0,0 +1,274 @@
+    use_displaced_mesh = true
+    compute_lm_residuals = false
+  []
+  [friction]
+    type = ApplyGitterleFrictionConstraintLMMechanicalContact
+    primary = 20
+    secondary = 10
+    variable = tangential_lm
+    primary_variable = disp_x
+    contact_pressure = normal_lm
+    gap_value = gap_variable
+    weighted_tangential_velocity = vel_variable
+    disp_y = disp_y
+    c_n = 1.0e3
+    c_t = 1.0e5
+    mu = 0.12
+  []
```
This is the object that computes PDASS tangential constraints proposed by Wohlmuth.

recuero commented on modules/contact/src/constraints/ApplyGitterleFrictionConstraintLMMechanicalContact.C:
```diff
@@ -0,0 +1,257 @@
+      LocalDN weighted_tangential_vel(_weighted_tangential_velocity);
+      //      weighted_tangential_vel.derivatives()[0] = v_dot_tan.derivatives()[0];
+      //      weighted_tangential_vel.derivatives()[1] = v_dot_tan.derivatives()[1];
+      weighted_tangential_vel.derivatives()[4] = 1.0;
+      LocalDN gap_value(_gap_value);
+
+      gap_value.derivatives()[0] = pinfo->_normal(0);
+      gap_value.derivatives()[1] = pinfo->_normal(1);
+      gap_value.derivatives()[3] = 1.0;
+
+      //      Moose::out << "*******************\n";
+      //      Moose::out << "Jvar is: " << jvar << "\n";
+      //      Moose::out << "_contact_pressure_id is: " << _contact_pressure_id << "\n";
+      //      Moose::out << "_primary_var.number() is: " << _primary_var.number() << "\n";
+      //      Moose::out << "_disp_y_id is: " << _disp_y_id << "\n";
+      //      Moose::out << "_weighted_tangential_velocity_id is: " <<
+      //      _weighted_tangential_velocity_id
+      //                 << "\n";
+      //      Moose::out << "_gap_value_id is: " << _gap_value_id << "\n";
+
+      auto term_1 = std::max(_mu * (dual_contact_pressure + _c_n * gap_value),
+                             std::abs(_u_secondary[_qp] + _c_t * weighted_tangential_vel * _dt)) *
+                    _u_secondary[_qp];
+      auto term_2 = _mu * std::max(0.0, dual_contact_pressure + _c_n * gap_value) *
+                    (_u_secondary[_qp] + _c_t * weighted_tangential_vel * _dt);
+
+      if (jvar == _contact_pressure_id)
+        return (term_1 - term_2).derivatives()[2];
+      else if (jvar == _primary_var.number())
+        return (term_1 - term_2).derivatives()[0];
+      else if (jvar == _disp_y_id)
+        return (term_1 - term_2).derivatives()[1];
+      else if (jvar == _gap_value_id)
+        return (term_1 - term_2).derivatives()[3];
+      else if (jvar == _weighted_tangential_velocity_id)
+        return (term_1 - term_2).derivatives()[4];
+    }
+  }
+  return 0.;
+}
```
Jacobian isn't great. Only considering SecondarySecondary so far.

recuero committed:
Add mortar frictional constraints.

Based on primal-dual active set strategy.

Refs. #17495

recuero committed:
Fix retrieval of weighted gap from friction class

recuero committed:
Add automatic differentiation API for neighbor velocities

recuero committed:
Add mortar friction documentation

recuero committed:
Add cylinder and bouncing block frictional tests.

These tests only run for a few steps. They can be run for longer and converted to examples.

recuero committed:
Add frictional ironing problem.

This problem is only set up to run for a few steps.

Together with Dewen Yushu @dewenyushu

Others: remove execution mask from "cylinder.e", fix test file doc name.

recuero commented on modules/contact/test/tests/pdass_problems/tests:
```diff
@@ -0,0 +1,42 @@
+    rel_err = 1.0e-4
+    min_parallel = 4
+    heavy = true
+    min_ad_size = '200'
```
Very oblique projections appear to be adding a significant number of variable dependencies from the primary surface.

dschwen replied:
I thought very oblique segments are not generated anymore?

recuero replied:
Extremely oblique segments that cause numerical problems aren't generated. Very oblique segments can make sense, and can span a number of primary surface's lower-d elements, thus a possible reason for the increased required ad size.

recuero replied:
So this test is just pie in the sky. Never gets tested.

dschwen commented on modules/contact/doc/content/source/constraints/ComputeFrictionalForceLMMechanicalContact.md:
```diff
@@ -0,0 +1,32 @@
+
+This object automatically enforces normal contact constraints by making calls to its parent class `ComputeWeightedGapLMMechanicalContact`, see [`ComputeWeightedGapLMMechanicalContact`](/ComputeWeightedGapLMMechanicalContact.md) for input parameters and details.
+
+It is preliminary recommended to select  `c` to be on the order of the moduli of elasticity of the bodies into contact, and `c_t` to be a few orders of magnitude less than `c`. This selection of these purely numerical parameters can represent an initial difficulty when running $new$ models, but they can be held constant once good convergence behavior has been attained.  
```
```suggestion
Preliminary recommendation is to select  `c` to be on the order of the moduli of elasticity of the bodies into contact, and `c_t` to be a few orders of magnitude less than `c`. This selection of these purely numerical parameters can represent an initial difficulty when running _new_ models, but they can be held constant once good convergence behavior has been attained.  
```

lindsayad commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -0,0 +1,215 @@
+  // Clear _node_to_weighted_gap map
+  ComputeWeightedGapLMMechanicalContact::residualSetup();
+
+  _node_to_weighted_tangential_velocity.clear();
```
Ok good! This was going to be my first hypothesis for why you were getting large sparsity patterns ... if you weren't clearing your containers and your mesh was moving a lot theoretically you could end up with a derivatives container holding the whole interface mesh! (Although then your Newton solves probably wouldn't have converged because your SMP Jacobian would have been so bad ... or wait actually they would have been fine ... just your residuals/physics wouldn't have been right :smile: )

lindsayad commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -0,0 +1,215 @@
+    for (_i = 0; _i < _test.size(); ++_i)
+      computeQpIProperties();
+  }
+}
```
I don't think this override is actually necessary right? This is what is already done in the base class?

lindsayad replied:
Oh good I thought I might be being stupid by just working on a single commit and i was right

lindsayad commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -0,0 +1,215 @@
+    _weighted_gap_ptr = &_node_to_weighted_gap[pr.first];
+    _tangential_vel_ptr = &pr.second;
+    enforceConstraintOnNode(pr.first);
+  }
```
So here you are going to end up doing two loops. One from `ComputeWeightedGapLMMechanicalContact::post` and then one from here. I don't think you should call `ComputeWeightedGapLMMechanicalContact::post` here. Instead you should uncomment the `Compute::WeightedGapLMMechanicalContact::enforceConstraintOnNode` call in your `enforceConstraintOnNode` override

recuero replied:
Makes sense

lindsayad commented on modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C:
```diff
@@ -111,7 +114,7 @@ ComputeWeightedGapLMMechanicalContact::residualSetup()
 void
 ComputeWeightedGapLMMechanicalContact::jacobianSetup()
 {
-  residualSetup();
+  ComputeWeightedGapLMMechanicalContact::residualSetup();
```
Haha this is a bit ridiculous but I understand why you must have it

lindsayad replied:
I changed my mind ... we shouldn't need this. You already override `jacobianSetup` in your derived and it calls `residualSetup` ... which will call the base `residualSetup` and then do your additional work in the derived `residualSetup`.

If a user, however, instantiates the base class then the `jacobianSetup` here will call the base `residualSetup` and everything will be hunky dory. So if somehow there *is* a need for this change, then something is wrong

lindsayad commented on modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C:
```diff
@@ -148,7 +151,7 @@ ComputeWeightedGapLMMechanicalContact::post()
   for (const auto & pr : _node_to_weighted_gap)
   {
     _weighted_gap_ptr = &pr.second;
-    enforceConstraintOnNode(pr.first);
+    ComputeWeightedGapLMMechanicalContact::enforceConstraintOnNode(pr.first);
```
If you don't call this in your derived then this shouldn't be necessary

lindsayad commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -0,0 +1,185 @@
+  residualSetup();
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::computeResidual(const Moose::MortarType mortar_type)
+{
+  if (mortar_type != Moose::MortarType::Lower)
+    return;
+
+  // Fill node to weighted tangential velocity map.
+  for (_qp = 0; _qp < _qrule_msm->n_points(); _qp++)
+  {
+    computeQpProperties();
+    for (_i = 0; _i < _test.size(); ++_i)
+      computeQpIProperties();
+  }
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::computeJacobian(const Moose::MortarType mortar_type)
+{
+  // During "computeResidual" and "computeJacobian" we are actually just computing properties on the
+  // mortar segment element mesh. We are *not* actually assembling into the residual/Jacobian. The
+  // property of interest is the map from node to weighted tangential velocity. Computation of the
+  // properties proceeds identically for residual and Jacobian evaluation hence why we simply call
+  // computeResidual here. We will assemble into the residual/Jacobian later from the post() method
+  computeResidual(mortar_type);
+}
```
Ok I'm reviewing all the commits now like a smart person. Yea I don't think we need these overrides

lindsayad commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -0,0 +1,185 @@
+  computeResidual(mortar_type);
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::post()
+{
+  // Enforce weighted gap constraint (calls enforceConstraintOnNode)
+  ComputeWeightedGapLMMechanicalContact::post();
+
+  // Enforce frictional complementarity constraints
+  for (const auto & pr : _node_to_weighted_tangential_velocity)
+  {
+    _weighted_gap_ptr = &_node_to_weighted_gap[pr.first];
+    _tangential_vel_ptr = &pr.second;
+    enforceConstraintOnNode(pr.first);
+  }
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::enforceConstraintOnNode(const Node * const node)
+{
+  // Get friction LM index
```
So my previous comments stand... don't call the base `post()` from derived `post()` but *do* call the base `enforceConstraintOnNode` from the derived `enforceConstraintOnNode`

recuero replied:
Yeah, the comments were clear and applied!

lindsayad reviewed (changes requested):
I think if we do this correctly, I should see no `ComputeWeightedGapLMMechanicalContact::` syntax in `ComputeWeightedGapLMMechanicalContact`

recuero committed:
Address Alex's comments on better hierarchy calling

lindsayad commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -0,0 +1,158 @@
+  InputParameters params = ComputeWeightedGapLMMechanicalContact::validParams();
+  params.addClassDescription("Computes the tangential frictional forces");
+  params.addRequiredCoupledVar("friction_lm", "The frictional Lagrange's multiplier");
+  params.addParam<Real>("c_t", 1e0, "Numerical parameter for tangential constraints");
```
I actually can't remember if we ended up bumping the default value for `c` for normal contact but I think I remember you thinking that we should. Do you think *this* default is appropriate?

recuero replied:
The few problems I have run converge well with low `c_t`. So, it is intended to be lower than `c` but maybe subject to revision in the future. 

You did increase the value of `c_n` in the base class to 1.0e6, which I think it's reasonable. Some authors recommend it to be on the order the Young's modulus so we may need to increase it when we deal with steel, we'll see.

recuero replied:
We can make  some sensitivity analysis on those parameters time permitting.

lindsayad commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -0,0 +1,158 @@
+    nodal_residual = friction_lm_value;
+  else
+  {
+    auto term_1 = std::max(_mu * (contact_pressure + _c * weighted_gap),
+                           std::abs(friction_lm_value + _c_t * tangential_vel * _dt)) *
+                  friction_lm_value;
+    auto term_2 = _mu * std::max(0.0, contact_pressure + _c * weighted_gap) *
+                  (friction_lm_value + _c_t * tangential_vel * _dt);
+
+    nodal_residual = term_1 - term_2;
```
I'm just going to assume that this math is right. You could make `term_1` and `term_2` `const auto` if you make any other changes.

recuero replied:
I will continue to test this class on new cases. Hopefully we'll catch any issues.

lindsayad commented on modules/contact/test/tests/bouncing-block-contact_pdass/try_gitterle.normal.i:
```diff
@@ -0,0 +1,267 @@
+starting_point = 2e-1
```
Oops, it looks like this file isn't tested in anyway

recuero replied:
Good catch. I always manage to forget removing all the input files I used for testing...

lindsayad reviewed (changes requested):
This looks really clean! I really like it. Just the untested file and now since you have to push again ayways you might as well make those variables `auto` -> `const auto`. I love `const`!

recuero committed:
Address Alex's comments II

recuero commented:
This new design that doesn't go through `NodeFaceConstraint` makes things much simpler to me. It makes having a decent Jacobian much easier and has a mortary feel to it 👍

lindsayad reviewed (approved):
This looks great to me! @dschwen let me know if you have any other feedback in the next day, otherwise I will merge

dschwen commented:
No, this looks good to me!

On Wed, May 26, 2021, 5:52 PM Alex Lindsay ***@***.***> wrote:

> ***@***.**** approved this pull request.
>
> This looks great to me! @dschwen <https://github.com/dschwen> let me know
> if you have any other feedback in the next day, otherwise I will merge
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/17497#pullrequestreview-669631985>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABRMPQBQSINEM6EOSMGWB3TPWCUVANCNFSM42JSTPDQ>
> .
>
